llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp

```diff
@@ -0,0 +1,142 @@
+//===- LowerESIMDVLoadVStore.cpp - lower vload/vstore to load/store -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// convert vload/vstore to load/store if they are not for genx_volatile.
+//
+// File scope simd variables marked with genx_volatile attribute want
+// guarranteed allocation in register file, therefore we use vload/vstore
+// instead of load/store, so they won't be optimized away by llvm.
+//
+// For ordinary simd variables, we do not need to protect load/store. But
+// there is no good way to do this in clang. So we need this pass in the
+// end of module passes to separate the cases that we need vload/vstore vs.
+// the cases that we do not need vload/vstore
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "loweresimdvloadvstore"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/GenXIntrinsics/GenXIntrinsics.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Scalar.h"
+
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeESIMDLowerLoadStorePass(PassRegistry &);
+}
+
+namespace {
+
+struct ESIMDLowerLoadStore : public FunctionPass {
+  static char ID;
+  ESIMDLowerLoadStore() : FunctionPass(ID) {
+    initializeESIMDLowerLoadStorePass(*PassRegistry::getPassRegistry());
+  }
+  virtual void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+  }
+
+  virtual bool runOnFunction(Function &F) override;
+
+private:
+  bool lowerLoadStore(Function &F);
+};
+
+} // namespace
+
+char ESIMDLowerLoadStore::ID = 0;
+INITIALIZE_PASS(ESIMDLowerLoadStore, "ESIMDLowerLoadStore",
+                "Lower ESIMD reference loads and stores", false, false)
+
+bool ESIMDLowerLoadStore::runOnFunction(Function &F) {
+  bool Changed = false;
+  Changed |= lowerLoadStore(F);
```

Comment:
Is it possible to replace these 3 lines with
return lowerLoadStore(F);  ?

Reply:
Why. It works and it is correct. Do we really need to pick on this detail?

Reply:
Is this review only about functional correctness?

Reply:
That is fine. I can change this one.